### PR TITLE
Jonghoon/predictive heads

### PR DESCRIPTION
**Pull request by sjhpark**
## Context
This code implements a tunable (switchable) classification head for a pose model.
The task for training a tunable head is named "pose-trainablehead".
It has new classes defined for task model, head, loss criterion, and trainer (e.g., "PoseTrainableHeadModel" for task model class).
This model passes in a one-hot vector of features (index of image source as of now) as the extra input argument to the classification head of the model.

image

image

## Test
1. Create a docker container for model training inside `~/verdant/dev/src/python/trainers/yolo8`.
- `./start.sh {data dir} {runs dir where checkpoints and ONNX files will be saved} shell -u {local ultralytics dir} -c {name of docker container to create}`
Example:
`./start.sh ~/data/lettuce_brassica_endive_tomato_artichoke_rose_011525/ ~/data/runs pose-tunablehead shell -u ~/verdant/ultralytics -c teset_031025`
2. Run `python simple_training.py` inside the docker container.
3. Once the training is complete and the ONNX format of the model is saved (e.g., "best.onnx"), double check whether it allows the user to pass in a one-hot vector as the 2nd input argument.
- Notes: The one-hot vector input must match the number of training sources (N). For example, with 5 sources, to indicate source index 2 (index starts from 0), use [0, 0, 1, 0, 0].

### CODE REVIEW

File: ultralytics/nn/modules/head.py [hunk 1]
```diff
@@ -312,6 +384,59 @@ def kpts_decode(self, bs, kpts):
...
+
+    def __init__(self, nc=80, kpt_shape=(17, 3), ch=()):
+        """Initialize YOLO network with default parameters and Convolutional Layers."""
+        super().__init__(nc, ch)
+        self.kpt_shape = kpt_shape  # number of keypoints, number of dims (2 for x,y or 3 for x,y,visible)
+        self.nk = kpt_shape[0] * kpt_shape[1]  # number of keypoints total
+        self.detect = DetectTunableHead.forward
+
+        c4 = max(ch[0] // 4, self.nk)
+        self.cv4 = nn.ModuleList(nn.Sequential(Conv(x, c4, 3), Conv(c4, c4, 3), nn.Conv2d(c4, self.nk, 1)) for x in ch)
```
Comment (romainthibaux) [inline ultralytics/nn/modules/head.py:400]: This could be implemented more simply by:
- inheriting from `PoseHead`
- Calling `super().__init__(nc, kpt_shape=kpt_shape, ch=[c + nc/2 for c in ch])`
- Leave `forward()` mostly as written, where the feature vector is expanded, concatenated to the input at each scale, and passed to the parent's `forward()`.

This would mean that the features also affect the bounding box side of the head as well, and that we would still train when the input feature vector is all zeros, but that's probably ok.

File: ultralytics/nn/tasks.py [hunk 2]
```diff
@@ -322,6 +344,142 @@ def __init__(self, cfg='yolov8n-pose.yaml', ch=3, nc=None, data_kpt_shape=(None,
...
+
+        Returns:
+            (torch.Tensor): The output of the network.
+        """
+        if isinstance(x, dict):  # case: training / validator
+            if self.training:
+                return self.loss(x, *args, **kwargs)
+            else:
+                x_features = self.convert_cls_to_features(x)  # convert class labels to one hot feature vectors
+                return self.predict([x['img'], x_features], *args, **kwargs)
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/tasks.py:450]: I just fixed it to use 2 arguments instead of a list
`return self.predict(x['img'], x_features, *args, **kwargs)`

However, I just noticed that since I am using PoseValidator instead of creating a new class for a validator, the image tensor x['img'] that I pass into the model's forward function during validation is not going through this PoseTunableHeadModel's forward function, but it is passed into PoseModel's forward function. Therefore, current validation approach is not accurate as PoseModel doesn't pass in x_features (one hot vector) into the model.
I need to create a new Validator class for my tunable head class by inheriting from PoseValidator. Example:
```
class PoseTunableHeadValidator(PoseValidator):
    def __init__(self, dataloader=None, save_dir=None, pbar=None, args=None, _callbacks=None):
        """Initialize a 'PoseTunableHeadValidator' object with custom parameters and assigned attributes."""
        super().__init__(self, dataloader=None, save_dir=None, pbar=None, args=None, _callbacks=None))
        self.args.task = 'pose-tunablehead'
```

File: ultralytics/engine/exporter.py [hunk 3]
```diff
@@ -211,9 +213,12 @@ def __call__(self, model=None):
 
         y = None
         for _ in range(2):
+            if isinstance(m, PoseTunableHeadModel):
+                y = model(im, image_features)  # take one hot vector of image features as extra input
             y = model(im)  # dry runs
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/engine/exporter.py:218]: That's right. I put `y = model(im)` inside `else`.
```
if isinstance(m, PoseTunableHeadModel):
    y = model(im, image_features)  # take one hot vector of image features as extra input
else:
    y = model(im)  # dry runs
```

File: ultralytics/nn/autobackend.py [hunk 4]
```diff
@@ -337,11 +337,7 @@ def forward(self, im, augment=False, visualize=False):
         Returns:
             (tuple): Tuple containing the raw output tensor, and processed output for visualization (if visualize=True)
         """
-        b, ch, h, w = im.shape  # batch, channel, height, width
-        if self.fp16 and im.dtype != torch.float16:
-            im = im.half()  # to FP16
-        if self.nhwc:
-            im = im.permute(0, 2, 3, 1)  # torch BCHW to numpy BHWC shape(1,320,192,3)
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/autobackend.py:344]: I have put them back.

File: ultralytics/nn/tasks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/tasks.py:450]: Update:
Per our discussion, I am not passing in a tuple (image, one hot vector) to the model during validation.
Instead of copying `__call__` function of `BaseValidator` and revising how inputs are passed into the model's forward function, I created a child class `PoseTunableHeadValidator` and overwrote preprocess to make `batch['img']` as a tuple `(batch['img'], one hot vector)`.
This change also required me to revise `AutoBackend` class (which is a model wrapper class for validation) in `autobackend.py` accordingly to handle the case when its input argument `im` is a tuple, not an image tensor.

File: ultralytics/nn/autobackend.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/autobackend.py:344]: Update:
Now, it is a function and is called for two if cases: im is an image tensor or a tuple.
```
def prepare_image(im):
    if self.fp16 and im.dtype != torch.float16:
        im = im.half()  # to FP16
    if self.nhwc:
        im = im.permute(0, 2, 3, 1)  # torch BCHW to numpy BHWC shape(1,320,192,3)
    return im
if isinstance(im, tuple):
    im, im_features = im
    im = prepare_image(im)
    is_im_tuple = True
else:
    im = prepare_image(im)
    is_im_tuple = False
```

File: ultralytics/nn/modules/head.py [hunk 5]
```diff
@@ -312,6 +384,59 @@ def kpts_decode(self, bs, kpts):
...
+        # reshape one hot feature vector (1D) to concatenate with input image tensor (x)
+        device = x[0].device
+        x_features_reshaped = []
+        for i in range(self.nl):  # per detection scale
+            B, _, H, W = x[i].shape  # different (H, W) shape per detection scale (e.g., 24x24, 48x48, 96x96)
+            x_features_reshaped.append(x_features.view(B, -1, 1, 1).expand(B, -1, H, W).to(device))  # (B, nc//2, H, W)
+
+        # concatenate x with one hot feature vector
+        x_with_features = [torch.concat((xi, x_features_reshaped_i), dim=1) for 
+                           (xi, x_features_reshaped_i) in zip(x, x_features_reshaped)]  # (B, C+nc//2, H, W)
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:416]: Per suggestion, I revised `PoseTunableHead` and `DtectTinableHead` to expand the feature vector and concatenate it with the image tensor across 3 different detection scales.

File: ultralytics/nn/modules/head.py (hunk 1 above)
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:400]: There are two options provided by you for me to choose.
1. This
2. https://github.com/Verdant-Robotics/ultralytics/pull/22#discussion_r1996435722
I chose the option #2 for now to not pass in the one-hot feature vector to the bbox regression path in the head.

**Review by romainthibaux — APPROVED** [review]

File: ultralytics/engine/validator.py [hunk 6]
```diff
@@ -159,6 +159,7 @@ def __call__(self, trainer=None, model=None):
         for batch_i, batch in enumerate(bar):
             self.run_callbacks('on_val_batch_start')
             self.batch_i = batch_i
+
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/engine/validator.py:162]: Deleted per suggestion.

File: ultralytics/models/yolo/pose/val.py [hunk 7]
```diff
@@ -213,3 +213,43 @@ def eval_json(self, stats):
...
+        args = dict(model='yolov8n-pose-tunablehead.pt', data='coco8-pose.yaml')
+        validator = PoseTunableHeadValidator(args=args)
+        validator()
+        ```
+    """
+
+    def __init__(self, dataloader=None, save_dir=None, pbar=None, args=None, _callbacks=None):
+        """Initialize a 'PoseTunableHeadValidator' object with custom parameters and assigned attributes."""
+        super().__init__(dataloader, save_dir, pbar, args, _callbacks)
+        self.sigma = None
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/models/yolo/pose/val.py:235]: Removed `sigma` and `kpt_shape`.

File: ultralytics/nn/autobackend.py [hunk 8]
```diff
@@ -337,11 +337,18 @@ def forward(self, im, augment=False, visualize=False):
...
+        def prepare_image(im):
+            if self.fp16 and im.dtype != torch.float16:
+                im = im.half()  # to FP16
+            if self.nhwc:
+                im = im.permute(0, 2, 3, 1)  # torch BCHW to numpy BHWC shape(1,320,192,3)
+            return im
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/autobackend.py:345]: Added an empty line after the function definition.
```
def prepare_image(im):

    if self.fp16 and im.dtype != torch.float16:
        im = im.half()  # to FP16
    if self.nhwc:
        im = im.permute(0, 2, 3, 1)  # torch BCHW to numpy BHWC shape(1,320,192,3)
    return im
```

File: ultralytics/nn/modules/head.py [hunk 9]
```diff
@@ -271,7 +350,7 @@ def forward(self, x):
         if self.training:
             return x, mc, p
         return (torch.cat([x, mc], 1), p) if self.export else (torch.cat([x[0], mc], 1), (x[1], mc, p))
-
+
```
[the remark this answers is not in the document]
Comment (sjhpark) [reply ultralytics/nn/modules/head.py:353]: Deleted a line.
